### DISCUSSION TITLE
TLS handshake failed. How to adjust tls_version_client_min option ?

### DISCUSSION THREAD

**Discussion by gg4u** [mitmproxy/mitmproxy#5489] [Q&A]
hi!

I got this error:

[::1]:57484: Client TLS handshake failed. Client and mitmproxy cannot agree on a TLS version to use. You may need to adjust mitmproxy's tls_version_client_min option.

How to adjust tls_version_client_min option ?

I also got : 

Client TLS handshake failed. The client disconnected during the handshake. If this happens consistently for keyvalueservice.icloud.com, this may indicate that the client does not trust the proxy's certificate.

Any workaround ?

--- EDIT -another question

Please note that I got the above errors while dumping only specific requests that belonged to a different domain (not icloud.com):

why did I get them?
do they prevent other requests?

e.g.  what I am doing is:
`mitmdump -w dump_me '~u notiCloud*'`

**Comment by meitinger** [comment]
> How to adjust tls_version_client_min option ?

You can try adding `--set tls_version_client_min=SSL3` to your command line options (or use the "Edit Options" button in the "Options" tab on mitmweb). The (current) default minimum is TLS1_2, so setting it to SSL3 will enable TLS1_1 and SSL3.

> Client TLS handshake failed. The client disconnected during the handshake. If this happens consistently for keyvalueservice.icloud.com, this may indicate that the client does not trust the proxy's certificate.

Have you installed mitmproxy's CA certificate on the client? If not see https://docs.mitmproxy.org/stable/concepts-certificates/ for details.

**Reply by gg4u** [reply]
Hi @meitinger , thanks for your answer:

> The (current) default minimum is TLS1_2, so setting it to SSL3 will enable TLS1_1 and SSL3.

for my general culture, could you briefly explain to a novice what TLS does and the difference between TLS1_2 and SSL3, so as to understand why TLS SSL3 would solve the issue of disconnecting ?

> have you installed mitmproxy's CA certificate on the client? If not see https://docs.mitmproxy.org/stable/concepts-certificates/ for details.

I installed the mitmproxy certificate authority.

About custom certificates, I read that : 
`Using a directory allows certs to be selected based on hostname, while using a filename allows a single specific certificate to be used for all TLS connections.`

could you explain a use case for using custom certificates ?


-- Edit

Unfortunately:

`mitmdump -w test6 "~u filter" --set tls_version_client_min=SSL3`

did not worked out.
```
...
::1]:64888: client connect
[::1]:64888: server connect keyvalueservice.icloud.com:443 (17.248.248.39:443)
[::1]:64888: Client TLS handshake failed. Client and mitmproxy cannot agree on a TLS version to use. You may need to adjust mitmproxy's tls_version_client_min option.
[::1]:64888: client disconnect
[::1]:64888: server disconnect keyvalueservice.icloud.com:443 (17.248.248.39:443)
[::1]:64894: client connect
[::1]:64894: server connect keyvalueservice.icloud.com:443 (17.248.248.39:443)
[::1]:64894: Client TLS handshake failed. The client disconnected during the handshake. If this happens consistently for keyvalueservice.icloud.com, this may indicate that the client does not trust the proxy's certificate.
...
```

**Reply by meitinger** [reply]
> for my general culture, could you briefly explain to a novice what TLS does and the difference between TLS1_2 and SSL3, so as to understand why TLS SSL3 would solve the issue of disconnecting ?

SSL and TLS are both protocols that enable encrypted ("secured") communication between two parties. TLS is the successor to SSL. The differences between the first version of TLS, i.e. TLS 1.0, and SSL are actually quite subtle. But one of the differences - that also continues to be different in higher version of TLS - is the list of ciphers that can or cannot be used to encrypt the data or hash certificates. TLS 1.3, for example, doesn't allow MD5, as it's no longer considered secure. This, together with other weaknesses found in the protocol itself, is the reason why implementers may decide to drop support for lower versions. If the client only supports older versions and none of the newer versions advertised by the server, the connection fails. This can be solved by forcing the server to accept lower versions as well, which in case of mitmproxy, is the mentioned command line option.

However, that is not the only reason a handshake may fail.

> About custom certificates, I read that : `Using a directory allows certs to be selected based on hostname, while using a filename allows a single specific certificate to be used for all TLS connections.`
> 
> could you explain a use case for using custom certificates ?

This section concerns client certificates: Here mitmproxy, acting as a client connecting to an upstream server, provides a certificate to the upstream server. Since it is quite likely that different servers require different certificates from their clients, one can achieve this by using the directory variant and storing the required certificate based on the hostname. The filename variant would provide the same certificate to all upstream servers.

Judging from the logs you provided, the upstream connection works fine, so custom client certificates shouldn't be necessary.

If you've installed the mitmproxy's CA certificate and this error persists, be advised that are other methods a client can use to detect whether mitmproxy's certificate is "fake" or not. You can read more about it in the [Certificate Pinning](https://docs.mitmproxy.org/stable/concepts-certificates/#certificate-pinning) section.

**Reply by gg4u** [reply]
`If you've installed the mitmproxy's CA certificate and this error persists, be advised that are other methods a client can use to detect whether mitmproxy's certificate is "fake" or not. `

I read the certificate pinning section:

keyvalueservice.icloud.com:443 is an Apple domain.
I tried to the --ignore-hosts option:

mitmdump -ignore-hosts "keyvalueservice.icloud.com:443" but did not have effect.

If relevant, since it is possible to read who issued the certificate: 

<img width="547" alt="Screenshot 2022-07-26 at 10 56 40 AM" src="https://user-images.githubusercontent.com/284214/181085251-f8bebb90-ab6a-4b15-b0aa-8254f0096dcb.png">


- might it be possible that a service may refuse to issue requests to a client with a mitmproxy certificate ?
- I don't know well how certificates works, but might it be possible to customize the name ?

**Reply by meitinger** [reply]
> mitmdump -ignore-hosts "keyvalueservice.icloud.com:443" but did not have effect.

Actually in your case the hostname is `keyvalueservice.icloud.com`, without the port. Here's the relevant part of the `ignore-host` usage description:
```
    The supplied value is interpreted as a regular expression and matched on the ip or the hostname.
```
Keep that in mind when specifying the value.

> might it be possible that a service may refuse to issue requests to a client with a mitmproxy certificate

Given your logs, it is the client who might refuse the certificate. You can try one of the tools listed in the [Certificate Pinning](https://docs.mitmproxy.org/stable/concepts-certificates/#certificate-pinning) section.

> I don't know well how certificates works, but might it be possible to customize the name ?

You can provide your own certificate using the `--cert` option. Otherwise you'll have to change the source. Keep in mind, however, that if the client uses certificate pinning, simply changing the name won't circumvent that.

**Reply by vishnuintellect** [reply]
> --set tls_version_client_min=SSL3

This is what has worked for me on iOS, started mitmproxy from terminal with this option and it worked.. Thanks @meitinger
